Added support for dynamic imposter ports

jgitter opened the pull request:
Made a number of changes to support dynamic clients:

1. Made port a nullable int? parameter on imposter constructors that defaults to null
2. Made port a nullable int? parameter on the factory methods exposed on the MB client
3. Updated the MountebankRequestProxy to set the Imposter port after submit for create requests
4. Added unit test cases for imposter constructors and mountebank client factory methods
5. Added a unit test for the request proxy to ensure that port is stripped from response content
6. Added an acceptance test verifying that the port can be null and is properly set  on the imposter after create.
7. Added example to the tutorial class for creating an imposter with no port.

Thanks again for building this client!

Commit:
Made a number of changes to support dynamic clients:

1. Made port a nullable int? parameter on imposter constructors that defaults to null
2. Made port a nullable int? parameter on the factory methods exposed on the MB client
3. Updated the MountebankRequestProxy to set the Imposter port after submit for create requests
4. Added unit test cases for imposter constructors and mountebank client factory methods
5. Added a unit test for the request proxy to ensure that port is stripped from response content
6. Added an acceptance test verifying that the port can be null and is properly set  on the imposter after create.
7. Added example to the tutorial class for creating an imposter with no port.

mattherman commented:
Thanks for the PR. I'll take a look at this as soon as I can. Also, ignore the failing build for now. I had been messing around with running the acceptance tests on AppVeyor and forgot to delete that build, so it failed.

jgitter commented:
Thanks for the quick response.  Let me know if you feel any additional tests are needed.

mattherman commented on MbDotNet/MountebankRequestProxy.cs:
```diff
@@ -121,6 +123,24 @@ private HttpResponseMessage ExecutePost(string resource, string json)
+                try
+                {
+                    var content = response.Content.ReadAsStringAsync().Result;
+                    Match match = new Regex(".*\"port\": (\\d+),", RegexOptions.Multiline).Match(content);
```
I with there was a better way to do this than just parsing through the response content. My first thought was to deserialize the JSON to a `RetrievedImposter` type like the `GetImposter` method does. Technically, we could since we have the original imposter object so we would know what protocol it is, but that feels weird as well. I think I am fine with this for now. I can always update it later if I think of a better way.

mattherman replied:
If you have any other ideas, I'd love to hear them.

jgitter replied:
I also considered adding the Port property to the RetrievedImposter and getting it that way, but that detail isn't really important for imposter verification which is the intended purpose of that model.  I could even add a new model type that only contains the port that is used here.  That might arguably be cleaner.

mattherman replied:
I think I would lean towards adding the new model. That way we can let the JSON deserializer do the work for us instead of parsing the content. Make the type internal though so it is not exposed in the package. I would be fine with just putting that class in this file even since this is the only place it will be used.

mattherman commented on MbDotNet/MountebankRequestProxy.cs:
```diff
@@ -121,6 +123,24 @@ private HttpResponseMessage ExecutePost(string resource, string json)
+                {
+                    var content = response.Content.ReadAsStringAsync().Result;
+                    Match match = new Regex(".*\"port\": (\\d+),", RegexOptions.Multiline).Match(content);
+                    imposter.SetDynamicPort(Convert.ToInt32(match.Groups[1].Value));
```
I think we should use `int.Parse` here instead of `Convert.ToInt32`. That way it will throw an exception instead of silently setting the port to zero.

jgitter replied:
Agreed, thanks for the suggestion.  I think this is a symptom of me being new to c#.

mattherman commented on MbDotNet/IClient.cs:
```diff
@@ -1,103 +1,103 @@
+        /// Creates a new imposter on the specified port with the HTTP protocol. The Submit method
+        /// must be called on the client in order to submit the imposter to mountebank.
+        /// </summary>
+        /// <param name="port">The port the imposter will be set up to receive requests on</param>
```
Could you please update the comments on the affected methods to signify what will happen if they do not provide a value for the port?

jgitter replied:
Will do.

mattherman commented:
Take a quick look at my comments. Other than that, I think it looks good. I might not publish a new version of the package with this change right away though, since making `Imposter.Port` nullable is technically a breaking change.

mattherman commented:
Also, since I can't figure out how to get AppVeyor to re-run the build against your PR, I ran everything locally with your changes. All the tests and CA looks good.

jgitter commented:
Is making Port nullable really a breaking change?  Anyone using an older version would need to have set a port value when creating their imposters.  My intention was that any code written by consumers of this library would be compatible now that port is nullable.  Is there a case where this is not true?

mattherman commented:
The updates to the `CreateXXXImposter` methods are backwards-compatible. It is the `Imposter.Port` nullability that is not. If a consumer of the package is using that property for anything, they would not expect it to possibly be null.

For example, with our changes, the following would no longer compile since `Port` is no longer of type `int`:
```
public void Main()
{
    var client = new MountebankClient();
    var imposter = client.CreateHttpImposter(1234);
    DoSomethingWithPort(imposter.Port);
}
public void DoSomethingWithPort(int port)
{
    ........
}

jgitter commented:
You're right.  I hadn't considered it since that would be such an unusual usage pattern.  However, it is technically possible.  In any case, the sooner an official build can be released, the better.  I will likely get to your review comments tomorrow.  Thanks!

mattherman commented:
I was trying to think of another way we could do this and have it be backwards-compatible. The only other option I thought of was to keep it a regular `int` and default it to some value like -1. However, with the current way I am serializing the imposters, I'm not sure how to get it to skip serializing the port in that case.

jgitter commented:
That pattern occurred to me as well, but I have a strong negative reflex against anything resembling a "magic value" like that -1.  Knowledge of that magic value leaks into the client factory methods, the models, and the request proxy.  I decided the nullable integer was more appropriate because it has an inherent semantic meaning, i.e. not set.

mattherman commented:
Agreed. We'll just have to accept the breaking change then.

Commit:
Added comments to IClient detailing the behavior of the nullable port.

Commit:
Added a new response model that contains only the port.  This is used by the request proxy for the purpose of retrieving the port number to set on the imposter object that was submitted.

jgitter commented:
I added your requested comments and added a new model for the imposter create response for use in the request proxy.  Have another look when you have a moment.

Thanks!

mattherman commented:
Looks good. Thanks for contributing!

jgitter commented:
Thank you for being so responsive!
